rust/arrow/src/compute/kernels/cast.rs

```diff
@@ -376,6 +378,27 @@ pub fn cast(array: &ArrayRef, to_type: &DataType) -> Result<ArrayRef> {
             Int64 => cast_string_to_numeric::<Int64Type>(array),
             Float32 => cast_string_to_numeric::<Float32Type>(array),
             Float64 => cast_string_to_numeric::<Float64Type>(array),
+            Date32(DateUnit::Day) => {
+                use chrono::{NaiveDate, NaiveTime};
+                let zero_time = NaiveTime::from_hms(0, 0, 0);
+                let string_array = array.as_any().downcast_ref::<StringArray>().unwrap();
+                let mut builder = PrimitiveBuilder::<Date32Type>::new(string_array.len());
+                for i in 0..string_array.len() {
+                    if string_array.is_null(i) {
+                        builder.append_null()?;
+                    } else {
+                        match NaiveDate::parse_from_str(string_array.value(i), "%Y-%m-%d")
+                        {
+                            Ok(date) => builder.append_value(
+                                (date.and_time(zero_time).timestamp() / SECONDS_IN_DAY)
+                                    as i32,
+                            )?,
+                            Err(_) => builder.append_null()?, // not a valid date
```

Comment:
I know this has been merged but do you really want to append null and not inform user when a value cannot be correctly parsed? This is the behaviour of Spark but is very dangerous in some cases (especially dealing with money).

Reply:
I was thinking about that as well, but this is a pattern repeated throughout the cast kernel so I thought it best to follow it; what do you think would be better to do instead?

Reply:
@andygrove this is a fundamental question about ANSI type SQL support vs a lower compatibility.

Apache Spark has put a lot of work into adding ANSI behaviour to Spark (https://spark.apache.org/docs/3.0.0/sql-ref-ansi-compliance.html#type-conversion) to correct error on invalid values which I believe is the correct approach rather than the error suppression strategy which Spark inherited from Hive SQL.

This is not the place for this discussion but it feels like a good time to address this kind of fundamental question.

Reply:
Thanks @seddonm1 this is a great point and I am actually pretty familiar with Spark's CAST and ANSI CAST support and I agree that Spark has a lot of dangerous behavior that hides data errors. I wasn't really thinking about this while reviewing this PR. 

I filed https://issues.apache.org/jira/browse/ARROW-10793 to faciliate this discussion.

Reply:
Thanks Andy. I'm keen to be more involved with DataFusion.

Reply:
A possible solution is to add cast options (there's a JIRA open somewhere), which we can use to tweak the behaviour. We could look at the options in the C++ implementations.
One cast option would be whether failed should return errors or be converted to nulls.

Reply:
@nevi-me 
Yes, that is definitely an option but I would hope that the default is strict with the option to relax for users that understand the risk.

We wrote a lot of code (https://github.com/tripl-ai/arc/blob/master/src/main/scala/ai/tripl/arc/transform/TypingTransform.scala) to safely apply correct data typing from strings in Spark in a way that allows us to parse the entire dataset and collect errors per row before returning. 

This works in one of two ways (like you have described):
- a way that returns NULL for invalid conversions and list of errors for the row (which we should have done a `map[field name, error]`) which can then be addressed by the user. If any field that contains an error is not nullable then the job fails immediately.
- a ways that fails fast simulating the ANSI approach of first error found throws error (which can be very slow to debug).